contrib/lib-autopilot.py

```diff
@@ -273,17 +276,87 @@ def get_candidates(self,k=10):
+            res_pdf[k] = (w * v + (1-w)/len(candidates))
+        return res_pdf
+    
+    def __calculate_proposed_channel_capacities(self,pdf, balance = 1000000):
+        minimal_channel_balance = 20000 #lnd uses 20k satoshi which seems reasonble
+        
+        min_probability = min(pdf.values())
+        needed_total_balance = math.ceil(minimal_channel_balance / min_probability)
+        self.__logger.info("Need at least a balance of {} satoshi to open {} channels".format(needed_total_balance,len(candidates)))
+        while needed_total_balance > balance: 
+            min_val = min(pdf.values())
+            k = [k for k, v in pdf.items() if v == min_val][0]
+            self.__logger.info("Not enough balance to open {} channels. Remove node: {} and rebalance pdf for channel balances".format(len(pdf),k))
+            del pdf[k]
+            
+            s = sum(pdf.values())
+            pdf = {k:v/s for k,v in pdf.items()}
             
+            min_probability = min(pdf.values())
+            needed_total_balance = math.ceil(minimal_channel_balance / min_probability)
+            self.__logger.info("Need at least a balance of {} satoshi to open {} channels".format(needed_total_balance,len(pdf)))
+                    
+        return pdf
+        
+        
+    
+    def connect(self,candidates, balance = 1000000):
+        pdf = self.__calculate_statistics(candidates)
+        connection_dict = self.__calculate_proposed_channel_capacities(pdf,balance)
+        for nodeid, fraction in connection_dict.items():
+            try:
+                satoshis = math.ceil(balance * fraction)
+                self.__logger.info("Try to open channel with a capacity of {} to node {}".format(satoshis,nodeid))
+                self.__rpc_interface.fundchannel(nodeid, satoshis)
+            except ValueError as e:
+                self.__logger.info("Could not open a channel to {} with capacity of {}. Error: {}".format(nodeid,satoshis,str(e)))
+        
     
-    def run(self):
+    def find_candidates(self,k=10):
         self.__logger.info("running the autopilot on a graph with {} nodes and {} edges.".format(len(self.G.nodes()), len(self.G.edges())))
-        candidates = self.get_candidates(k=21)
+        candidates = self.get_candidates(k)
         time.sleep(1)
-        self.__display_nodes(list(candidates))
-        
-
 
+        #FIXME: Temprorary for coding so that candidates don't have to be calculated all the time
+        #with open("data/candidates","wb") as outfile:
+        #    pickle.dump(candidates,outfile,pickle.HIGHEST_PROTOCOL)
+        return candidates
 
 if __name__ == '__main__':
     autopilot = Autopilot()
-    autopilot.run()
\ No newline at end of file
+    candidates = autopilot.find_candidates(21)
+    #FIXME: Temporary for debugging
+    #with open("data/candidates","rb") as f:
+    #    candidates = pickle.load(f)
+    autopilot.connect(candidates,1000000)
```

Comment:
Might want to check with `lightningd` before attempting to allocate a fixed amount of satoshis.

Reply:
Also we probably don't want to connect to 20 candidates every time we run this, rather we should be attempting to find `20 - current_number_of_peers` candidates and connect to those.

Reply:
the question is really how should it be integrated to c-lighting? I imagined also via the lightning-cli so something like `lightning-cli runautopilot` there could be a mandatory argument of how many channels the node wants to be connected to. also it could reading a value from the config file. I guess there would also be other config variables. Which process do you suggest?